### PR TITLE
Build type nodeselectors for worker builds

### PR DESCRIPTION
**Pull request by rcerven**
Signed-off-by: Robert Cerven <rcerven@redhat.com>

### CODE REVIEW

File: osbs/api.py [hunk 1]
```diff
@@ -602,11 +612,18 @@ def create_worker_build(self, **kwargs):
         for required in ('platform', 'release', 'arrangement_version'):
             if not kwargs.get(required):
                 missing.add(required)
-
         if missing:
             raise ValueError("Worker build missing required parameters: %s" %
                              missing)
 
+        extra = set()
+        for unwanted in ('platforms'):
```
Comment (vrutkovs) [inline osbs/api.py]: We're interating over a tuple with single item, why?

Comment (rcerven) [inline osbs/api.py]: well it is prepared for future changes if there would be more unwanted things, and same as in orchestrator

File: osbs/api.py [hunk 2]
```diff
@@ -644,6 +661,14 @@ def create_orchestrator_build(self, **kwargs):
             raise OsbsValidationException("can't create orchestrate build "
                                           "when can_orchestrate isn't enabled")
 
+        extra = set()
+        for unwanted in ('platform', 'release'):
+            if kwargs.get(unwanted):
+                extra.add(unwanted)
+        if extra:
```
Comment (vrutkovs) [inline osbs/api.py]: `extra = [x for x in kwargs.keys() if x in ['platform', 'release']]` ?

**Comment by bfontecc007** [comment]
LGTM!

File: osbs/api.py (hunk 2 above)
Comment (rcerven) [inline osbs/api.py]: I can't use that because the keys are defined but None

File: osbs/api.py [hunk 3]
```diff
@@ -395,13 +393,21 @@ def _create_build_config_and_build(self, build_request):
             logger.debug('Changed parent ImageStreamTag? %s', changed_ist)
 
         if triggers:
-            existing_bc['spec']['triggers'] = triggers
-            self.os.update_build_config(build_config_name, json.dumps(existing_bc))
```
Comment (lcarva) [inline osbs/api.py]: Is this change needed? I'd rather not mess around with this method.

File: osbs/api.py [hunk 4]
```diff
@@ -395,13 +393,21 @@ def _create_build_config_and_build(self, build_request):
...
+            # so in order to completely replace 'triggers',
+            # we have to first empty it by setting it 'null'
+            # and after we can add triggers
+            patch_triggers = {'spec': {'triggers': None}}
+            self.os.patch_build_config(build_config_name, json.dumps(patch_triggers))
+            patch_triggers = {'spec': {'triggers': triggers}}
+            self.os.patch_build_config(build_config_name, json.dumps(patch_triggers))
 
         if image_stream and triggers:
-            prev_version = existing_bc['status']['lastVersion']
```
Comment (lcarva) [inline osbs/api.py]: Is this change needed? I'd rather not mess around with this method.

File: osbs/api.py [hunk 5]
```diff
@@ -602,11 +612,13 @@ def create_worker_build(self, **kwargs):
         for required in ('platform', 'release', 'arrangement_version'):
             if not kwargs.get(required):
                 missing.add(required)
-
         if missing:
             raise ValueError("Worker build missing required parameters: %s" %
                              missing)
 
+        if kwargs.get('platforms'):
```
Comment (lcarva) [inline osbs/api.py]: I agree with this change, but it seems unrelated to node selectors. Can we split up into different commits?

File: osbs/api.py [hunk 6]
```diff
@@ -643,6 +655,13 @@ def create_orchestrator_build(self, **kwargs):
         if not self.can_orchestrate():
             raise OsbsValidationException("can't create orchestrate build "
                                           "when can_orchestrate isn't enabled")
+        extra = set()
+        for unwanted in ('platform', 'release'):
```
Comment (lcarva) [inline osbs/api.py]: I agree with this change, but it seems unrelated to node selectors. Can we split up into different commits?

File: osbs/build/build_request.py [hunk 7]
```diff
@@ -1186,7 +1195,22 @@ def render(self, validate=True):
         if koji_task_id is not None:
             self.set_label('koji-task-id', str(koji_task_id))
 
-        self.template['spec']['nodeSelector'] = self.platform_node_selector
+        # for worker builds set nodeselectors
```
Comment (lcarva) [inline osbs/build/build_request.py]: Let's move this bonanza to its own method, maybe `render_node_selectors`.

File: osbs/build/build_request.py [hunk 8]
```diff
@@ -1186,7 +1195,22 @@ def render(self, validate=True):
...
+        if self.spec.platforms.value is None:
```
Comment (lcarva) [inline osbs/build/build_request.py]: I'd rather this was a check for `platform`, not `platforms`. There's also a `build_type` spec value now, too. It's probably better to use `build_type` because we really are checking the build type here. Checking `platform` or `platforms` is just a heuristic for achieving that.

File: osbs/build/build_request.py [hunk 9]
```diff
@@ -1186,7 +1195,22 @@ def render(self, validate=True):
...
+                self.template['spec']['nodeSelector'] = self.auto_build_node_selector
+            else:
+                self.template['spec']['nodeSelector'] = self.explicit_build_node_selector
+
+            # platform nodeselector
+            if self.platform_node_selector:
+                self.template['spec']['nodeSelector'].update(self.platform_node_selector)
+
+            # scratch build nodeselector
+            if self.scratch:
```
Comment (lcarva) [inline osbs/build/build_request.py]: Would this add both `scratch_build_node_selector` and either `explicit_build_node_selector` or `auto_build_node_selector`.
Be definition, you can't have a scratch autorebuild. Maybe this:

```python
if self.is_scratch:
  ... # scratch build
elif self.is_auto:
  ... # auto build
else:
  ... # explicit build

File: osbs/api.py (hunk 4 above)
Comment (rcerven) [inline osbs/api.py]: it isn't needed, as the lastVersion is still same

File: osbs/build/build_request.py (hunk 8 above)
Comment (rcerven) [inline osbs/build/build_request.py]: platform isn't passed to method only platforms, 'build_type' ??? where is it?

File: osbs/build/build_request.py (hunk 9 above)
Comment (rcerven) [inline osbs/build/build_request.py]: that would add only scratch and not 'scratch & explicit'

yeah I know that by definition you can't have that

File: osbs/build/build_request.py (hunk 8 above)
Comment (rcerven) [inline osbs/build/build_request.py]: and we have already our logic based on platforms

File: osbs/build/build_request.py (hunk 7 above)
Comment (rcerven) [inline osbs/build/build_request.py]: moved to it's own method

File: osbs/api.py (hunk 6 above)
Comment (rcerven) [inline osbs/api.py]: split into 2 commits

File: osbs/api.py (hunk 3 above)
Comment (rcerven) [inline osbs/api.py]: well the code isn't needed, because we are using patch now

Comment (lcarva) [inline osbs/api.py]: Let's remove this change then.

File: osbs/api.py (hunk 4 above)
Comment (lcarva) [inline osbs/api.py]: Let's remove this change then.

File: osbs/api.py (hunk 3 above)
Comment (lcarva) [inline osbs/api.py]: TBH I don't see how switching to patch makes this better. In either case, it's completely irrelevant to this PR.

File: osbs/core.py [hunk 10]
```diff
@@ -200,6 +200,10 @@ def _put(self, url, with_auth=True, **kwargs):
         headers, kwargs = self._request_args(with_auth, **kwargs)
         return self._con.put(url, headers=headers, verify_ssl=self.verify_ssl, **kwargs)
 
+    def _patch(self, url, with_auth=True, **kwargs):
```
Comment (vrutkovs) [inline osbs/core.py]: Tests don't hit those lines

File: osbs/core.py [hunk 11]
```diff
@@ -373,6 +377,13 @@ def update_build_config(self, build_config_id, build_config_json):
         check_response(response)
         return response
 
+    def patch_build_config(self, build_config_id, build_config_json):
```
Comment (vrutkovs) [inline osbs/core.py]: Tests don't hit those lines

File: osbs/api.py (hunk 3 above)
Comment (vrutkovs) [inline osbs/api.py]: Right, there already is `clean_triggers` in `utils.py`.

The comment should be added why are triggers being replaced at all - it seems pointless here

Comment (rcerven) [inline osbs/api.py]: clean triggers is cleaning not ImageChange triggers from existing config

this lines were suppose to update buildconfig with new triggers

File: osbs/conf.py [hunk 12]
```diff
@@ -554,3 +543,36 @@ def get_platform_node_selector(self, platform):
                 nodeselector = dict([k.strip(), v.strip()] for (k, v) in raw_nodeselector.items())
 
         return nodeselector
+
+    def get_scratch_build_node_selector(self):
+        nodeselector = {}
+        nodeselector_str = self._get_value("scratch_build_node_selector", self.conf_section,
+                                           "scratch_build_node_selector")
+        if nodeselector_str and nodeselector_str != 'none':
```
Comment (vrutkovs) [inline osbs/conf.py]: This piece of code is being copy-pasted across three functions

File: tests/build_/test_build_request.py [hunk 13]
```diff
@@ -2105,3 +2105,70 @@ def test_info_url_format(self):
...
+                'explicit2': 'yes'
+            },
+            'scratch': {
+                'scratch1': 'yes',
+                'scratch2': 'yes'
+            }
+        }
+        nodeselector_expected = {}
+        if not platforms:
+            if platform:
```
Comment (vrutkovs) [inline tests/build_/test_build_request.py]: This is repeating the logic of the app. Instead `nodeselector_expected` should be passed as a param to the function

File: osbs/conf.py (hunk 12 above)
Comment (rcerven) [inline osbs/conf.py]: yeah as many other methods in conf.py, so I thought is better keep it nice and separate

File: tests/build_/test_build_request.py (hunk 13 above)
Comment (rcerven) [inline tests/build_/test_build_request.py]: so I should write manually all combinations?

Comment (vrutkovs) [inline tests/build_/test_build_request.py]: yes, it would be easier to review

File: osbs/conf.py (hunk 12 above)
Comment (vrutkovs) [inline osbs/conf.py]: Other methods in `conf.py` are one liners, this one is 7 lines. Whenever the format of the selectors needs to be changed 3 functions would require an update

File: osbs/build/build_request.py (hunk 9 above)
Comment (lcarva) [inline osbs/build/build_request.py]: Then let's change the code to reflect that.

File: osbs/conf.py (hunk 12 above)
Comment (lcarva) [inline osbs/conf.py]: +1 for making a little helper method to handle parsing of the raw value.

File: tests/build_/test_arrangements.py [hunk 14]
```diff
@@ -510,6 +510,7 @@ class TestArrangementV3(TestArrangementV2):
                 'add_labels_in_dockerfile',
                 'koji_parent',
                 'reactor_config',
+                'check_and_set_rebuild',
```
Comment (lcarva) [inline tests/build_/test_arrangements.py]: See the comment a few lines above: `# Changing this? Add test methods`

File: osbs/conf.py [hunk 15]
```diff
@@ -547,22 +536,50 @@ def get_equal_labels(self):
...
+    def get_scratch_build_node_selector(self):
+        nodeselector_str = self._get_value("scratch_build_node_selector", self.conf_section,
+                                           "scratch_build_node_selector")
+        nodeselector = self.generate_nodeselector_dict(nodeselector_str)
+        return nodeselector
+
+    def get_explicit_build_node_selector(self):
+        nodeselector_str = self._get_value("explicit_build_node_selector", self.conf_section,
+                                           "explicit_build_node_selector")
+        nodeselector = self.generate_nodeselector_dict(nodeselector_str)
```
Comment (vrutkovs) [inline osbs/conf.py]: Nitpick: `return self.generate_nodeselector_dict(nodeselector_str)`?

Comment (rcerven) [inline osbs/conf.py]: changed

File: osbs/build/build_request.py [hunk 16]
```diff
@@ -1138,6 +1147,23 @@ def render_customizations(self):
...
+        if self.spec.platforms.value is None:
+
+            # auto or explicit build selector
+            if self.is_auto:
+                self.template['spec']['nodeSelector'] = self.auto_build_node_selector
+            else:
+                self.template['spec']['nodeSelector'] = self.explicit_build_node_selector
+                # scratch build nodeselector
+                if self.scratch:
+                    self.template['spec']['nodeSelector'].update(self.scratch_build_node_selector)
```
Comment (lcarva) [inline osbs/build/build_request.py]: why do we want scratch and explicit node selectors for the same build?
The intent was to have either one of the three types set, not a combination of some.

File: tests/test_conf.py [hunk 17]
```diff
@@ -492,3 +493,28 @@ def test_get_registry_api_versions(self, platform, config, expected, valid):
...
+        ({},
+         {}),
+        ({'default': {}},
+         {}),
+        ({'default': {'node_selector': 'breakfast=eggs.com'}},
+         {'breakfast': 'eggs.com'}),
+        ({'default': {'node_selector': 'breakfast=eggs.com, lunch=ham.com'}},
+         {'breakfast': 'eggs.com', 'lunch': 'ham.com'}),
+    ])
+    def test_get_lowprio_node_selectors(self, nodeselector_type, config, expected):
```
Comment (lcarva) [inline tests/test_conf.py]: This method should be renamed.

File: osbs/build/build_request.py [hunk 18]
```diff
@@ -1138,6 +1147,24 @@ def render_customizations(self):
...
         self.dj.dock_json_set_param('client_version', client_version)
 
+    def render_node_selectors(self):
+        # for worker builds set nodeselectors
+        if self.spec.platforms.value is None:
+
+            # auto or explicit build selector
+            if self.is_auto:
+                self.template['spec']['nodeSelector'] = self.auto_build_node_selector
+            else:
```
Comment (lcarva) [inline osbs/build/build_request.py]: This could be an `elif self.scratch` to reduce the nested statements.

**Comment by lcarva** [comment]
Updated release notes.
